Change XMLHttpRequest error handling

gnandretta opened the pull request:
- Listen for 'load' event instead of 'readystatechange' because the
  later can be triggered along with the 'error' event.
- Check the response status code, and invoke the callback with an
  error argument when it is not in the 20x range.

gnandretta committed:
Change XMLHttpRequest error handling

- Listen for 'load' event instead of 'readystatechange' because the
  later can be triggered along with the 'error' event.
- Check the response status code, and invoke the callback with an
  error argument when it is not in the 20x range.

cristiandouce commented:
So, this avoids the `fn` to be called twice when there is an error, right? @gnandretta

gnandretta commented:
@cristiandouce yes, and it will also invoke the callback with an error argument when the response doesn't succeed, for instance when a 404 is returned.

gjohnson commented on index.js:
```diff
@@ -62,13 +62,15 @@ function json(url, obj, headers, fn){
-    if (4 == req.readyState) return fn(null, req);
+    // sometimes IE returns 1223 when it should be 204
+    // see http://stackoverflow.com/questions/10046972/msie-returns-status-code-of-1223-for-ajax-request
+    /^(20\d|1223)$/.test(req.status) ? fn(null, req) : fn(req);
```
Shouldn't the falsey case here create an error to bubble up?

gnandretta replied:
@gjohnson Sure! I've also created the error for the call in the `onerror` handler as well. See commit 13ccd0de0fbd159b7b38073a096f61cf68df4316

gnandretta committed:
Bubble up actual Error instances

gnandretta committed:
Update callback error arguments

- Rename error.evt to error.event for the network error case.
- Pass the request as the second argument instead of as an event
  property for the unsuccessful request case.

gnandretta commented:
@gjohnson any chance on get this merged?
